api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -77,82 +77,107 @@ class Stats:
-    obs_tables_iter: somacore.ReadIter[pa.Table]
-    """Iterates the SOMA batches (tables) of corresponding ``obs`` and ``X`` data. This is an internal class, 
+class _ObsAndXSOMAIterator(Iterator[_SOMAChunk]):
+    """Iterates the SOMA chunks of corresponding ``obs`` and ``X`` data. This is an internal class,
     not intended for public use."""
 
     X: soma.SparseNDArray
     """A handle to the full X data of the SOMA ``Experiment``"""
 
-    var_joinids: pa.Array
+    obs_joinids_chunks_iter: Iterator[npt.NDArray[np.int64]]
+
+    var_joinids: npt.NDArray[np.int64]
     """The ``var`` joinids to be retrieved from the SOMA ``Experiment``"""
 
-    def __init__(self, X: soma.SparseNDArray, obs_tables_iter: somacore.ReadIter[pa.Table], var_joinids: pa.Array):
-        self.obs_tables_iter = obs_tables_iter
+    def __init__(
+        self,
+        obs: soma.DataFrame,
+        X: soma.SparseNDArray,
+        obs_column_names: Sequence[str],
+        obs_joinids_chunked: npt.NDArray[np.int64],  # 2D
+        var_joinids: npt.NDArray[np.int64],
+        shuffle: bool = False,
+    ):
+        self.obs = obs
         self.X = X
+        self.obs_column_names = obs_column_names
+        self.obs_joinids_chunks_iter = self._maybe_local_shuffle_obs_joinids(obs_joinids_chunked, shuffle)
         self.var_joinids = var_joinids
 
-    def __next__(self) -> _ObsAndXSOMABatch:
-        pytorch_logger.debug("Retrieving next SOMA batch...")
+    @staticmethod
+    def _maybe_local_shuffle_obs_joinids(
+        obs_joinids_chunked: npt.NDArray[np.int64], shuffle: bool  # 2D
+    ) -> Iterator[npt.NDArray[np.int64]]:
+        rng = np.random.default_rng()
+        return (
+            rng.permutation(obs_joinid_chunk) if shuffle else obs_joinid_chunk
+            for obs_joinid_chunk in obs_joinids_chunked
+        )
+
+    def __next__(self) -> _SOMAChunk:
+        pytorch_logger.debug("Retrieving next SOMA chunk...")
         start_time = time()
 
         # If no more batches to iterate through, raise StopIteration, as all iterators do when at end
-        obs_table = next(self.obs_tables_iter)
-        obs_batch = obs_table.to_pandas()
+        obs_joinids_chunk = next(self.obs_joinids_chunks_iter)
+
+        obs_batch = (
+            self.obs.read(
+                coords=(obs_joinids_chunk,),
+                column_names=self.obs_column_names,
+            )
+            .concat()
+            .to_pandas()
+            .set_index("soma_joinid")
+        )
+        assert obs_batch.shape[0] == obs_joinids_chunk.shape[0]
 
         # handle case of empty result (first batch has 0 rows)
         if len(obs_batch) == 0:
             raise StopIteration
 
-        X_batch = _fast_csr.read_scipy_csr(self.X, obs_table["soma_joinid"].combine_chunks(), self.var_joinids)
+        # reorder obs rows to match obs_joinids_chunk ordering, which may be shuffled
+        obs_batch = obs_batch.reindex(obs_joinids_chunk, copy=False)
```

Comment:
N.B.

Reply:
I am reasonably certain that TileDB only sorts if you ask it to do so (and by default, it does not - see the `result_order` parameter). I believe what you are observing is that we _write_ the Census in row sorted order, and tiledbsoma is just reading them back in on-disk order (without a sort).

I don't think this distinction matters for your code - just clarifying for shared understanding.

Note: if you do care about order, you can explicitly ask for it using the above param.

Reply:
Ah, that makes sense, ty!